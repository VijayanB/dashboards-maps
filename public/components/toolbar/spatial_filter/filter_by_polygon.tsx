/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { useState } from 'react';
import { EuiPopover, EuiContextMenu, EuiPanel, EuiButtonIcon, EuiButton } from '@elastic/eui';
import { FilterInputPanel } from './filter_input_panel';
import polygon from '../../../images/polygon.svg';
import {
  DrawFilterProperties,
  DRAW_FILTER_POLYGON,
  DRAW_FILTER_POLYGON_DEFAULT_LABEL,
  DRAW_FILTER_SPATIAL_RELATIONS,
  DRAW_FILTER_SHAPE_TITLE,
} from '../../../../common';
import { FILTER_DRAW_MODE } from '../../../../common';

interface FilterByPolygonProps {
  setDrawFilterProperties: (properties: DrawFilterProperties) => void;
  isDrawActive: boolean;
}

export const FilterByPolygon = ({
  setDrawFilterProperties,
  isDrawActive,
}: FilterByPolygonProps) => {
  const [isPopoverOpen, setPopover] = useState(false);

  const onClick = () => {
    setPopover(!isPopoverOpen);
  };

  const closePopover = () => {
    setPopover(false);
  };

  const onSubmit = (input: { relation: string; label: string; mode: FILTER_DRAW_MODE }) => {
    setDrawFilterProperties({
      mode: input.mode,
      relation: input.relation,
      filterLabel: input.label,
    });
    closePopover();
  };

  const panels = [
    {
      id: 0,
      title: DRAW_FILTER_SHAPE_TITLE,
      content: (
        <FilterInputPanel
          drawLabel={DRAW_FILTER_POLYGON}
          defaultFilterLabel={DRAW_FILTER_POLYGON_DEFAULT_LABEL}
          relations={DRAW_FILTER_SPATIAL_RELATIONS}
          onSubmit={onSubmit}
          mode={FILTER_DRAW_MODE.POLYGON}
        />
      ),
    },
  ];

  const drawPolygonButton = (
    <EuiPanel paddingSize="none" className="spatialFilterToolbar__shape">
      <EuiButtonIcon
        color="text"
        size={'s'}
        iconType={polygon}
        onClick={onClick}
        aria-label={'draw_filter_polygon'}
        title={DRAW_FILTER_POLYGON}
        isDisabled={isDrawActive}
      />
    </EuiPanel>
  );
  return (
    <EuiPopover
      id="drawPolygonId"
      button={drawPolygonButton}
      isOpen={isPopoverOpen}
      closePopover={closePopover}
      panelPaddingSize="none"
      anchorPosition="leftUp"
      data-test-subj="drawPolygonPopOver"
    >
      <EuiContextMenu initialPanelId={0} panels={panels} />
    </EuiPopover>
  );
};
