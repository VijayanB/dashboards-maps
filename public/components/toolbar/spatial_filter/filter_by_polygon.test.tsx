/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */
import { shallow } from 'enzyme';
import React from 'react';
import { FilterByPolygon } from './filter_by_polygon';

it('renders filter by polygon button', () => {
  const mockCallback = jest.fn();
  const polygonComponent = shallow(
    <FilterByPolygon setDrawFilterProperties={mockCallback} isDrawActive={false} />
  );
  expect(polygonComponent).toMatchSnapshot();
});

it('renders filter by polygon in middle of drawing', () => {
  const mockCallback = jest.fn();
  const polygonComponent = shallow(
    <FilterByPolygon setDrawFilterProperties={mockCallback} isDrawActive={true} />
  );
  expect(polygonComponent).toMatchSnapshot();
});
