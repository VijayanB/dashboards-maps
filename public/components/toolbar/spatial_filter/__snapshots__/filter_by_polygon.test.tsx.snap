// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders filter by polygon button 1`] = `
<EuiPopover
  anchorPosition="leftUp"
  button={
    <EuiPanel
      className="spatialFilterToolbar__shape"
      paddingSize="none"
    >
      <EuiButtonIcon
        aria-label="draw_filter_polygon"
        color="text"
        iconType={Object {}}
        isDisabled={false}
        onClick={[Function]}
        size="s"
        title="Draw Polygon"
      />
    </EuiPanel>
  }
  closePopover={[Function]}
  data-test-subj="drawPolygonPopOver"
  display="inlineBlock"
  hasArrow={true}
  id="drawPolygonId"
  isOpen={false}
  ownFocus={true}
  panelPaddingSize="none"
>
  <EuiContextMenu
    initialPanelId={0}
    panels={
      Array [
        Object {
          "content": <FilterInputPanel
            defaultFilterLabel="polygon"
            drawLabel="Draw Polygon"
            mode="polygon"
            onSubmit={[Function]}
            relations={
              Array [
                "intersects",
                "disjoint",
                "within",
              ]
            }
          />,
          "id": 0,
          "title": "DRAW SHAPE",
        },
      ]
    }
    size="m"
  />
</EuiPopover>
`;

exports[`renders filter by polygon in middle of drawing 1`] = `
<EuiPopover
  anchorPosition="leftUp"
  button={
    <EuiPanel
      className="spatialFilterToolbar__shape"
      paddingSize="none"
    >
      <EuiButtonIcon
        aria-label="draw_filter_polygon"
        color="text"
        iconType={Object {}}
        isDisabled={true}
        onClick={[Function]}
        size="s"
        title="Draw Polygon"
      />
    </EuiPanel>
  }
  closePopover={[Function]}
  data-test-subj="drawPolygonPopOver"
  display="inlineBlock"
  hasArrow={true}
  id="drawPolygonId"
  isOpen={false}
  ownFocus={true}
  panelPaddingSize="none"
>
  <EuiContextMenu
    initialPanelId={0}
    panels={
      Array [
        Object {
          "content": <FilterInputPanel
            defaultFilterLabel="polygon"
            drawLabel="Draw Polygon"
            mode="polygon"
            onSubmit={[Function]}
            relations={
              Array [
                "intersects",
                "disjoint",
                "within",
              ]
            }
          />,
          "id": 0,
          "title": "DRAW SHAPE",
        },
      ]
    }
    size="m"
  />
</EuiPopover>
`;
